### ISSUE TITLE
[API][S2] regex constraint re-classifies the full vocabulary on every request, uncached

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#1568]
## Fact
- Constraint compilation runs inside the request's critical path, on the scheduler thread, and two of its caches do not cover the work they were built for: `prepare_regex()` has no reuse check at all and re-classifies the whole vocabulary on every request, and the JSON-Schema cache is defeated by a probe that re-parses the schema document before the cache is consulted.
- `src/runtime/engine_scheduler.cpp:1016`
- `src/runtime/constraint_manager.cpp:334`
- `src/runtime/constraint_manager.cpp:147`
- `src/runtime/engine.cpp:468`

## Evidence
- Compilation is synchronous on the prefill path:
```
$ sed -n '1008,1017p' src/runtime/engine_scheduler.cpp
    // Constraints via the per-request ConstraintManager. The old engine-global
    // manager was re-prepared here for EVERY prefill (constrained or not),
    // which clobbered the FSM of any concurrently decoding constrained
    // request. Prepare once on first need; later chunks reuse the state.
    // thinking_open = req->in_think_block: if the prompt already closed the
    // <think> block (e.g. /no_think emits an empty <think></think> in the
    // prompt), no </think> is ever generated — the preamble gate must enforce
    // immediately instead of absorbing prose until the budget.
    ensure_constraints_(req);
```

### A. Regex: full vocabulary re-classification on every request
```
$ sed -n '334,350p' src/runtime/constraint_manager.cpp
bool ConstraintManager::prepare_regex(const std::string& pattern, Tokenizer* tokenizer,
                                     bool thinking_open) {
    active_regex_ = false;
    active_json_ = false;
    active_schema_ = false;
    if (pattern.empty() || !tokenizer)
        return false;
    if (!regex_constrainer_)
        regex_constrainer_ = std::make_unique<RegexConstrainer>();
    // vocab_size here is the LOGITS width; the constrainer clamps its own
    // classification to the tokenizer vocab (SafeTensors models pad lm_head
    // past it — indexing to the logits width read out of bounds once).
    if (!regex_constrainer_->init(pattern, tokenizer)) {
```
  `init()` is unconditional — there is no `pattern == cached` guard. What it does:
```
$ sed -n '173,190p' src/compute/regex_constrain.cu
bool RegexConstrainer::init(const std::string& pattern, Tokenizer* tokenizer) {
    if (!tokenizer)
        return false;
    if (!init_pattern_only(pattern))
        return false;

    vocab_size_ = tokenizer->vocab_size();
    token_texts_.resize(vocab_size_);
    for (int i = 0; i < vocab_size_; i++)
        token_texts_[i] = tokenizer->decode_token(static_cast<int32_t>(i));
    eos_ids_ = tokenizer->eos_ids();

    if (!dev_.alloc_token_allow("RegexConstrainer", vocab_size_)) {
        initialized_ = false;
        return false;
    }
```
  plus `init_pattern_only` clearing the warm mask cache at `src/compute/regex_constrain.cu:116` (`mask_cache_.clear();`) and recompiling the NFA at `:126`.
- The sibling path shows this is an omission, not a design — `prepare_grammar` guards the identical work, and its comment states the cost:
```
$ sed -n '338,346p' src/runtime/constraint_manager.cpp
    // Same grammar this pooled manager already serves: skip re-compiling and
    // re-classifying ~151K tokens, and keep the warm mask cache — an agent that
    // pins one grammar sends it on every request.
    const bool reuse = grammar_constrainer_ && grammar_constrainer_->is_initialized() &&
                       grammar_constrainer_->source() == gbnf;
    if (!reuse) {
        if (!grammar_constrainer_)
            grammar_constrainer_ = std::make_unique<GrammarConstrainer>();
```
- The engine's pool key for regex is dead by construction. It is built as `"regex:" + pattern`:
```
$ sed -n '432,442p' src/runtime/engine.cpp
    const std::string pool_key_regex = regex_enforced ? ("regex:" + req->regex_pattern) : std::string();
    const std::string pool_key_grammar = grammar_enforced ? ("gbnf:" + req->grammar) : std::string();
    const std::string pool_key = tool_enforced ? ConstraintManager::tool_call_key(req->tool_constraint_tools,
                                                                                  req->tool_envelope_open,
                                                                                  req->tool_envelope_close,
                                                                                  req->tool_constraint_xml)
                                 : regex_enforced   ? pool_key_regex
                                 : grammar_enforced ? pool_key_grammar
                                                    : req->json_schema;
```
  but the lookup only ever compares against a schema key or a `"gbnf:"`-prefixed grammar key:
```
$ sed -n '464,478p' src/runtime/engine.cpp
std::shared_ptr<ConstraintManager> Engine::constraints_checkout_(const std::string& json_schema) {
    // Prefer a pooled manager that already classified this schema or grammar
    // (the key is the schema string, or "gbnf:" + the grammar source).
    if (!json_schema.empty()) {
        for (auto it = constraint_pool_.begin(); it != constraint_pool_.end(); ++it) {
            const std::string& cached_gbnf = (*it)->cached_grammar();
            if ((*it)->cached_schema() == json_schema ||
                (!cached_gbnf.empty() && json_schema == "gbnf:" + cached_gbnf)) {
                auto cm = std::move(*it);
                constraint_pool_.erase(it);
                return cm;
            }
        }
    }
```
  `cached_schema()` returns `cached_schema_string_`, which is only ever assigned a raw JSON-Schema document (`src/runtime/constraint_manager.cpp:168`) or a `tool-call:` key (`:281`) — never a `"regex:"`-prefixed string:
```
$ grep -n "cached_schema_string_" src/runtime/constraint_manager.cpp src/runtime/constraint_manager.h
src/runtime/constraint_manager.h:127:    const std::string& cached_schema() const noexcept { return cached_schema_string_; }
src/runtime/constraint_manager.h:139:    std::string cached_schema_string_;
src/runtime/constraint_manager.cpp:159:            json_schema == cached_schema_string_) {
src/runtime/constraint_manager.cpp:168:                    cached_schema_string_ = json_schema;
src/runtime/constraint_manager.cpp:175:                    cached_schema_string_.clear();
src/runtime/constraint_manager.cpp:273:    if (!(schema_constrainer_ && schema_constrainer_->is_initialized() && key == cached_schema_string_)) {
src/runtime/constraint_manager.cpp:278:            cached_schema_string_.clear();
src/runtime/constraint_manager.cpp:281:        cached_schema_string_ = key;
```
  so the regex branch of the pool lookup can never match.

### B. JSON Schema: the document is re-parsed before the cache is consulted
```
$ sed -n '145,166p' src/runtime/constraint_manager.cpp
    bool use_schema = !json_schema.empty();
    if (use_schema) {
        auto probe = parse_json_schema(json_schema);
        const SchemaNode* probe_res = probe ? resolve_schema_ref(probe.get(), probe.get()) : nullptr;
        if (probe_res && probe_res->type == SchemaType::OBJECT && probe_res->properties.empty() &&
            probe_res->enum_values.empty()) {
            IMP_LOG_INFO("ConstraintManager: free-form object schema → any-JSON constrainer");
            use_schema = false;
            json_mode = true;
        }
    }

    if (use_schema) {
        if (schema_constrainer_ && schema_constrainer_->is_initialized() &&
            json_schema == cached_schema_string_) {
            configure_gate(schema_constrainer_.get());
```
  `parse_json_schema` at `:147` runs unconditionally, ahead of the cache hit at `:158-159`, and it is not a cheap probe: it walks the whole document and recompiles every `pattern` into an NFA:
```
$ sed -n '514,519p' src/compute/json_schema.cpp
    if (!validate_refs(root.get(), root.get())) {
        IMP_LOG_ERROR("Failed to parse JSON schema: unresolvable $ref");
        return nullptr;
    }
    compile_patterns(root.get());
    return root;
```
  The cache therefore covers only the vocabulary classification in `SchemaConstrainer::init`, never the parse.

### C. Cache key shape (asked; not a collision)
- The schema/tool cache key is `cached_schema_string_`, a single field shared by two producers: the raw schema document (`:168`) and `tool_call_key()` (`:281`). `tool_call_key` is prefixed:
```
$ sed -n '67,74p' src/runtime/constraint_manager.h
    static std::string tool_call_key(const std::vector<std::pair<std::string, std::string>>& tools,
                                     const std::string& envelope_open, const std::string& envelope_close,
                                     bool xml = false) {
        std::string key = (xml ? "xml-tool-call:" : "tool-call:") + envelope_open + "\x1f" + envelope_close;
        for (auto& [name, params] : tools)
            key += "\x1f" + name + "\x1e" + params;
        return key;
    }
```
  A raw schema document can never equal a `tool-call:`-prefixed string, so the two producers cannot collide today. The key is a full-string compare, not a hash, so there is no hash-collision surface either. **This half is clean** — but it is clean by accident of the prefix, on a field with no type discriminator, which is what makes the regex-key bug above possible in the same structure.

## Impact
- A regex-constrained request pays a full vocabulary walk (`vocab_size` × `decode_token()` + `vocab_size` `std::string` constructions) plus an NFA compile plus a device allocation, on the scheduler thread, on **every** request — including the second and every subsequent request with the identical pattern. The comment at `constraint_manager.cpp:339-341` names the workload as "~151K tokens" and names the exact usage pattern it hurts: "an agent that pins one grammar sends it on every request". That is the `guided_regex` / `response_format:{"type":"regex"}` surface, `docs/API.md:64`.
- The warm mask cache is destroyed each time too (`src/compute/regex_constrain.cu:116`), so even the per-state masks are recomputed.
- Because it runs on the scheduler thread inside prefill, the cost is paid by every other in-flight request in the batch, not only by the constrained one.
- For `json_schema`, every request re-parses the schema and recompiles each `pattern` NFA even on a cache hit. Agent workloads that pin one schema pay it per turn.
- Blast radius: `/v1/chat/completions`, `/v1/messages`, `/v1/responses` with `regex_pattern` / `guided_regex` / `response_format:{"type":"regex"}` (case A) and with `response_format:{"type":"json_schema"}` (case B). All architectures and quants — this is host-side.

## Repro
```bash
# NOT RUN — needs a running server; measurement is the caller's to schedule.
# A: same regex, 20 sequential requests; TTFT should be flat after the first if cached.
for i in $(seq 1 20); do
  curl -s -o /dev/null -w '%{time_starttransfer}\n' localhost:8080/v1/chat/completions \
    -H 'content-type: application/json' \
    -d '{"model":"<id>","max_tokens":8,"temperature":0,
         "messages":[{"role":"user","content":"a phone number"}],
         "response_format":{"type":"regex","regex":"[0-9]{3}-[0-9]{4}"}}'
done
# Control arm — the grammar path, which IS cached (constraint_manager.cpp:341):
#   same loop with {"type":"grammar","grammar":"root ::= [0-9]{3} \"-\" [0-9]{4}"}
# A flat grammar arm against a non-flat regex arm isolates the missing guard
# from generic server noise.

# NOT RUN — B: same json_schema with a `pattern` field, 20 sequential requests,
# server log at diagnostics.log_level=debug; count parse_json_schema entries.
```

## Fix direction
- Give `prepare_regex` the same guard `prepare_grammar` has: keep the pattern on the constrainer (`RegexConstrainer::pattern_` already exists, `src/compute/regex_constrain.cu:118`) and skip `init()` when it matches and `is_initialized()`.
- Make the regex pool key reachable: either store `"regex:" + pattern` in a `cached_regex_string_` with its own accessor next to `cached_grammar()`, or replace the three ad-hoc string keys with one tagged key type so a new constraint kind cannot silently miss the lookup again.
- Move the free-form-object probe behind the cache check, or cache its verdict alongside `cached_schema_string_`, so a cache hit does not re-parse.
- Add a counter (the `#1321` lesson: a path with no observable counter cannot be tested) — e.g. `constraint_compiles_total` — so a regression here is measurable rather than inferred from latency.

## Acceptance criteria
- [ ] Two sequential requests with an identical `regex_pattern` call `RegexConstrainer::init` exactly once. Prove it with a counter or a debug log line count, not with a timing.
- [ ] Two sequential requests with an identical `json_schema` call `parse_json_schema` exactly once.
- [ ] `constraints_checkout_` returns a pooled manager for the second identical regex request; assert on the pool-hit path directly.
- [ ] A CPU-lane test covers all three; mutation-validated by removing the new guard.

## Unverified
- No millisecond figure is claimed. No committed file in this repo states the cost of a vocabulary classification, and this audit ran nothing. The repro above is what turns "uncached" into a number, and it is listed as a measurement request.

**Comment by kekzl** [comment]
Fixed in b4600d5d (#1711). Verified on main.
